src/components/shared/ButtonWithPanel/ConfirmDialog.js

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+
+import * as React from 'react';
+import classNames from 'classnames';
+
+import { ClosePanelContext } from './ClosePanelContext';
+
+import './ConfirmDialog.css';
+
+type Props = {|
+  // If these callbacks return promises, they will be waited for before the
+  // panel is closed. To make this explicit we added Promise<mixed> even if that
+  // isn't necessary to pass Flow checks.
+  +onCancelButtonClick?: () => mixed | Promise<mixed>,
+  +onConfirmButtonClick?: () => mixed | Promise<mixed>,
+  +className?: string,
+  +title: string,
+  +cancelButtonText?: string,
+  +confirmButtonText?: string,
+  +confirmButtonType?: 'default' | 'primary' | 'destructive',
+  +children: React.Node,
+|};
+
+type PropsWithClose = {|
+  ...Props,
+  closePanel: void => mixed,
+|};
+
+type State = {|
+  status: 'idle' | 'working',
+|};
+
+export class ConfirmDialogImpl extends React.PureComponent<
+  PropsWithClose,
+  State
+> {
+  state = {
+    status: 'idle',
+  };
+
+  _onConfirmButtonClick = async () => {
+    if (this.state.status === 'working') {
+      return;
```

Comment:
codecov warning here is because the buttons are disabled in this test, so we can't reach this if block. I still added these if condition in case we want to remove the "disabled" attribute later.